modules/ingester/instance_search.go

```diff
@@ -205,6 +243,106 @@ func (i *instance) searchWAL(ctx context.Context, req *tempopb.SearchRequest, p
+	span, ctx := opentracing.StartSpanFromContext(ctx, "instance.searchWALWithTraceQL")
+	defer span.Finish()
+
+	opts := common.DefaultSearchOptions()
+	engine := traceql.NewEngine()
+	metrics := &tempopb.SearchMetrics{}
+	var traces []*tempopb.TraceSearchMetadata
+	var blockErrs []error
+
+	// i.traces has live Traces
+	// TODO: searching live traces with TraceQL is not supported yet
+
+	// Lock blocks mutex until all block search is done. This avoids
+	// deadlocking with other activity (ingest, flushing), caused by releasing
+	// and then attempting to retake the lock.
+	i.blocksMtx.RLock()
+	defer i.blocksMtx.RUnlock()
+	span.LogFields(ot_log.Event("blocksMtx acquired"))
+
+	// search headBlock
+	res, err := engine.Execute(ctx, req, traceql.NewSpansetFetcherWrapper(func(ctx context.Context, req traceql.FetchSpansRequest) (traceql.FetchSpansResponse, error) {
+		// span.SetTag("blockID", i.headBlock.BlockMeta().BlockID.String())
+		return i.headBlock.Fetch(ctx, req, opts)
+	}))
+	if err != nil {
+		blockErrs = append(blockErrs, err)
+	}
+
+	// collect headBlock results
+	traces = append(traces, res.Traces...)
+	metrics = mergeSearchMetrics(metrics, res.Metrics)
+
+	span.LogFields(
+		ot_log.String("msg", "done searching headBlock"),
+		ot_log.Int("traces_size", len(traces)))
+
+	if len(traces) >= maxResults {
+		// hit limit, return early
+		return traces, metrics, blockErrs
+	}
+
+	// search completingBlocks
+	for _, block := range i.completingBlocks {
+		if len(traces) >= maxResults {
+			break // exit early
+		}
+
+		// TODO: Execute will parse the query again and again... maybe fix that
+		// TODO: we are building NewSpansetFetcherWrapper for each block, feels bit wasteful :)
+		r, err := engine.Execute(ctx, req, traceql.NewSpansetFetcherWrapper(func(ctx context.Context, req traceql.FetchSpansRequest) (traceql.FetchSpansResponse, error) {
+			return block.Fetch(ctx, req, opts)
+		}))
+
+		if err != nil {
+			blockErrs = append(blockErrs, err)
+			continue // collect error and continue to next block
+		}
+		traces = append(traces, r.Traces...)
+		metrics = mergeSearchMetrics(metrics, r.Metrics)
+	}
+	span.LogFields(
+		ot_log.String("msg", "done searching completingBlocks"),
+		ot_log.Int("traces_size", len(traces)))
+
+	// search completeBlocks
+	for _, block := range i.completeBlocks {
+		if len(traces) >= maxResults {
+			break // exit early
+		}
+		r, err := engine.Execute(ctx, req, traceql.NewSpansetFetcherWrapper(func(ctx context.Context, req traceql.FetchSpansRequest) (traceql.FetchSpansResponse, error) {
+			return block.Fetch(ctx, req, opts)
+		}))
+		if err != nil {
+			blockErrs = append(blockErrs, err)
+			continue // collect error and continue to next block
+		}
+		traces = append(traces, r.Traces...)
+		metrics = mergeSearchMetrics(metrics, res.Metrics)
+	}
+	span.LogFields(
+		ot_log.String("msg", "done searching completeBlocks"),
+		ot_log.Int("traces_size", len(traces)))
```

Comment:
Something like `total_traces_size` would be more clear that this is also includes traces from the `headBlock` and `completingBlocks`.

```suggestion
		ot_log.Int("total_traces_size", len(traces)))
```